### PR TITLE
Simplify IRankedMenu interface

### LINKED ISSUE #10921: Simplify IRankedMenu interface
## Description

In https://github.com/jupyterlab/jupyterlab/pull/10909, this line was changed:

https://github.com/jupyterlab/jupyterlab/pull/10909/files#diff-10d72a5eea33520df6a5801ea65b535d663f6992db9991d69e933d9cd491ecfeR16

```typescript
export interface IRankedMenu extends IDisposable {
```
became
```typescript
export interface IRankedMenu extends Menu {
```

and this change was released in JupyterLab 3.1.9.

This is a big expansion of that IRankedMenu interface from a very simple interface with only a few selected methods to a much larger inherited interface that exposes implementation details from the Menu class. I think this was done to expose the `.removeItem` and `.items` attributes of the menu in the interface, as needed in that PR.

In general, we have not typically had interfaces extend actual classes (in this case, the `IRankedMenu` was changed to inherit from the Lumino `Menu` class), because it puts the class implementation details into the interface, and means that future class changes automatically extend the interface. Usually we've tried to use the interface to present the data and operations we intentionally wanted to expose to the user, regardless of the underlying implementation.

In this particular case of the IRankedMenu interface, the general thoughts about menu design were that one extension should not be able to remove another extension's menu items, or in general change the menu structure, other than adding its own menu items. #10879 and #10909 exposed a new usecase for an extension needing to be able to change its own menu items that may have been specified declaratively in its setting schema. One approach is to say dynamic menu items like that can only be added in code (and you decide in code whether to add them or not). Another approach is to do what #10909 did and expose the removeItem method on the menu. In either case, I think we should be conservative with how we extend the IRankedMenu interface, and just expose the methods needed for this usecase.

Actually, I think I would prefer saying that dynamic menu items must be added in code (not via the settings schema). That seems cleaner than adding an item then removing it if we didn't want it. @fcollonval - what do you think? Is that a valid usecase for not using the declarative menu system?

### PR DESCRIPTION
**Pull request by fcollonval**
## References

Fixes #10921

## Code changes

Reduce `IRankedMenu` interface to forbid removal of item menu as implemented before 3.1. This means:
- `menu` attribute has been removed
- `IRankedMenu` does not extends Lumino `Menu`

A new class `DisposableMenuItem` has been created to allow the consumer to dispose a single item added by `IRankedMenu.addItem`

## User-facing changes

N/A

## Backwards-incompatible changes

Reduce the interface introduce in 3.1.

### CODE REVIEW

**Review by jasongrout — COMMENTED** [review]
Thanks! I've left a few comments. I think when we backport this to 3.1.x, we'll need to be careful to maintain compatibility with 3.1.0.

File: packages/ui-components/src/components/menu.ts [hunk 1]
```diff
@@ -25,23 +33,15 @@ export interface IRankedMenu extends Menu {
    *
    * @param options - The options for creating the menu item.
    *
-   * @returns The menu item added to the menu.
+   * @returns The disposable menu item added to the menu.
    */
-  addItem(options: IRankedMenu.IItemOptions): Menu.IItem;
+  addItem(options: IRankedMenu.IItemOptions): IDisposable;
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply packages/ui-components/src/components/menu.ts:38]: I propose to use `IDisposable` to reflect the signature of `addGroup`.

Thanks for the note for the backport.

File: packages/terminal-extension/src/index.ts [hunk 2]
```diff
@@ -179,6 +172,43 @@ function activate(
   addCommands(app, tracker, settingRegistry, translator, options);
 
   if (mainMenu) {
+    // Add "Terminal Theme" menu below "JupyterLab Themes" menu.
+    const themeMenu = new Menu({ commands });
+    themeMenu.title.label = trans.__('Terminal Theme');
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply packages/terminal-extension/src/index.ts:177]: Great suggestion

**Review by blink1073 — APPROVED** [review]
Thank you!

**Comment by fcollonval** [comment]
@meeseeksdev please backport to 3.1.x

**Pull request merged**